### DISCUSSION TITLE
Indexing Performance

### DISCUSSION THREAD

**Discussion by rickysarraf** [photoprism/photoprism#2287] [General]
So, every morning, I run a docker command to process the new photos/videos that may have been added to my `originals` folder.

I do it with:

```
Call PhotoPrism Index and cleanup                                                                                                                   
time="2022-04-27T05:08:00Z" level=info msg="indexing originals in /photoprism/originals/Processed"          
time="2022-04-27T05:08:00Z" level=info msg="classify: loading nasnet"                                                                               
2022-04-27 05:08:00.585932: I tensorflow/cc/saved_model/reader.cc:31] Reading SavedModel from: /opt/photoprism/assets/nasnet                
2022-04-27 05:08:01.196756: I tensorflow/cc/saved_model/reader.cc:54] Reading meta graph with tags { photoprism }
2022-04-27 05:08:01.645179: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not 
compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA                                                                                                         
2022-04-27 05:08:03.263930: I tensorflow/cc/saved_model/loader.cc:202] Restoring SavedModel bundle.                     
2022-04-27 05:08:10.436722: I tensorflow/cc/saved_model/loader.cc:311] SavedModel load for tags { photoprism }; Status: success. Took 9851091 microseconds.                                                                                                                                             
time="2022-04-27T05:08:11Z" level=info msg="classify: loading labels from labels.txt"                              
time="2022-04-27T05:08:37Z" level=info msg="media: PXL_20220426_102004026.MP.jpg was taken at 2022-04-26 04:50:04 +0000 UTC (meta)"
time="2022-04-27T05:08:37Z" level=info msg="media: PXL_20220426_124056742.jpg was taken at 2022-04-26 07:10:56 +0000 UTC (meta)"
time="2022-04-27T05:09:06Z" level=info msg="index: matched 2 labels with PXL_20220426_124056742.jpg [29.312938736s]"
time="2022-04-27T05:09:06Z" level=info msg="nsfw: loading nsfw"
2022-04-27 05:09:06.694057: I tensorflow/cc/saved_model/reader.cc:31] Reading SavedModel from: /opt/photoprism/assets/nsfw
2022-04-27 05:09:06.748718: I tensorflow/cc/saved_model/reader.cc:54] Reading meta graph with tags { serve }
2022-04-27 05:09:06.930525: I tensorflow/cc/saved_model/loader.cc:202] Restoring SavedModel bundle.
time="2022-04-27T05:09:07Z" level=info msg="index: matched 2 labels with PXL_20220426_124048767.MP.jpg [29.978486052s]"
2022-04-27 05:09:08.339285: I tensorflow/cc/saved_model/loader.cc:311] SavedModel load for tags { serve }; Status: success. Took 1645222 microseconds.
time="2022-04-27T05:09:08Z" level=info msg="nsfw: loading labels from labels.txt"

...snipped...

time="2022-04-27T05:09:18Z" level=info msg="index: stacked related jpg file photoprism/storage/sidecar/Processed/PXL_20220426_102004026.MP.mp4.jpg"
time="2022-04-27T05:09:21Z" level=info msg="index: stacked related jpg file Processed/PXL_20220426_124344747.mp4.jpg"
time="2022-04-27T11:26:41Z" level=info msg="index: found no orphan files [202.093995ms]"
time="2022-04-27T11:26:46Z" level=info msg="indexed 315 files in 6h18m48.283780305s"
time="2022-04-27T11:26:46Z" level=info msg="closed database connection"
```

What surprises me the most is that it took 6+ hours to index just half a dozen of new files, from a total of just 315 files, in the `/Processed` folder. This is the same amount of time it would take to index the entire `originals`.

The exact command in use is: `docker exec photoprism_photoprism_1 photoprism index /Processed`.
Interestingly, if I instead, go to the webui and run `Start Indexing` there, it is almost just around 30 mins.

Is the `index` command, that is invoked through the webui, any different than the CLI ?

**Comment by srett** [comment]
I'm doing pretty much the exact same thing and never ran into this issue so far (~38k photos). I guess a first step would be increasing the log level to debug, to hopefully reveal what exactly is taking so long.

**Reply by lastzero** [reply]
It might run additional tasks after indexing, although I believe this is optional. I suspect an issue with recent database migrations, maybe a new index on the files table involving the storage path and missing flag. Check GitHub for related reports! We have friends visiting, so can't help much right now.

**Reply by rickysarraf** [reply]
@srett I've enabled debug logs and am stuck in the same scenario as yesterday. I'm impatiently waiting for the command run to complete, hoping to see some extra messages. But so far, nothing interesting has been logged.


@lastzero I think that is the case. After the indexing, `facial updates` may be being run and that too on the entire collection. But I don't know if there's an easy way to validate that. Right now, my logs are stuck at:

```
time="2022-04-28T07:06:10Z" level=debug msg="media: created no new thumbnails for PXL_20220427_142828542 [330.709µs]"
time="2022-04-28T07:06:10Z" level=debug msg="search: updated partial file index [105.543154ms]"
time="2022-04-28T07:06:11Z" level=debug msg="faces: found no orphan markers [289.631664ms]"
time="2022-04-28T07:06:11Z" level=debug msg="markers: found no invalid references [700.111409ms]"
time="2022-04-28T07:06:11Z" level=debug msg="markers: found no missing subjects [2.456775ms]"
```

I don't think that the additional tasks that may be run are `optional`.  because the command being run is just: `docker exec photoprism_photoprism_1 photoprism index /Processed`. So it is only picking the default for the `index` command

**Reply by srett** [reply]
Are you running with SQLite? If yes then this might be a missing index, if using MariaDB, most likely not, since then the high load should be generated by its process. (According to your `ps aux` output in the other comment)

**Reply by rickysarraf** [reply]
No. I'm on mariadb.

```
rrs@lenovo:/media/SSHD/.Pictures$ ps aux | grep mysql
999       763933  0.2  3.2 3542344 265088 ?      Ssl  11:52   1:15 mysqld --transaction-isolation=READ-COMMITTED --character-set-server=utf8mb4 --collation-server=utf8mb4_unicode_ci --max-connections=512 --innodb-rollback-on-timeout=OFF --innodb-lock-wait-timeout=120 --innodb-buffer-pool-size=2G
```

On start, I did get these migration messages, but I believe there's nothing wrong in them:

```
time="2022-04-28T07:04:49Z" level=debug msg="config: overriding config with values from /photoprism/storage/config/options.yml"                                                                   
time="2022-04-28T07:04:49Z" level=debug msg="config: running on 'Intel(R) Core(TM) i7-4510U CPU @ 2.00GHz', 8.3 GB memory detected"                                                               
time="2022-04-28T07:04:49Z" level=debug msg="settings: loaded from /photoprism/storage/config/settings.yml "                                                                                      
time="2022-04-28T07:04:49Z" level=debug msg="config: successfully initialized [12.872866ms]"                                                                                                      
time="2022-04-28T07:04:51Z" level=debug msg="migrate: found 17 previous migrations"                                                                                                               
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20211121-094727 skipped"                                                                                                                    
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20211124-120008 skipped"                                                                                                                    
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20220329-030000 skipped"                                                                                                                    
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20220329-040000 skipped"                                                                                                                    
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20220329-050000 skipped"                                                                                                                    
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20220329-060000 skipped"                                                                                                                    
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20220329-061000 skipped"                                                                                                                    
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20220329-070000 skipped"                                                                                                                    
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20220329-071000 skipped"                                                                                                                    
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20220329-080000 skipped"                                                                                                                    
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20220329-081000 skipped"                                                                                                                    
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20220329-082000 skipped"                                                                                                                    
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20220329-090000 skipped"                                                                                                                    
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20220329-091000 skipped"                                                                                                                    
time="2022-04-28T07:04:51Z" level=debug msg="migrate: 20220329-092000 skipped"                                                                                                                    
time="2022-04-28T07:04:52Z" level=debug msg="migrate: completed in 2.460415073s"                                                                                                                  
time="2022-04-28T07:04:52Z" level=info msg="indexing originals in /photoprism/originals/Processed"                                                                                                
time="2022-04-28T07:04:52Z" level=info msg="classify: loading nasnet"                            
2022-04-28 07:04:52.343746: I tensorflow/cc/saved_model/reader.cc:31] Reading SavedModel from: /opt/photoprism/assets/nasnet                                                                      
2022-04-28 07:04:53.111397: I tensorflow/cc/saved_model/reader.cc:54] Reading meta graph with tags { photoprism }                                                                                 
2022-04-28 07:04:53.616566: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2022-04-28 07:04:55.264060: I tensorflow/cc/saved_model/loader.cc:202] Restoring SavedModel bundle.                                                                                               
2022-04-28 07:05:02.182076: I tensorflow/cc/saved_model/loader.cc:311] SavedModel load for tags { photoprism }; Status: success. Took 9838374 microseconds.
time="2022-04-28T07:05:03Z" level=info msg="classify: loading labels from labels.txt"                                                                                                             
time="2022-04-28T07:05:04Z" level=debug msg="exiftool: extracting metadata from Processed/20220426_123246.jpg"                                                                                    
time="2022-04-28T07:05:04Z" level=debug msg="exiftool: extracting metadata from Processed/20220426_123141.jpg"                                                                                    
time="2022-04-28T07:05:04Z" level=debug msg="exiftool: extracting metadata from Processed/20220427_165749.jpg"                                                                                    
time="2022-04-28T07:05:04Z" level=debug msg="exiftool: extracting metadata from Processed/20220427_165747.jpg"                                                                                    
time="2022-04-28T07:05:05Z" level=debug msg="index: created ff39a17d476523e3e69bf9ebb63966dc84f4113e_exiftool.json"                                                                               
time="2022-04-28T07:05:05Z" level=debug msg="index: created fc57d4b88f3e5d7fb1f40eb46ece1c23590b325b_exiftool.json"                                                                               

... snipped ...

```

**Reply by rickysarraf** [reply]
The job completed just now.

```
time="2022-04-28T14:50:35Z" level=debug msg="faces: found no ambiguous subjects [7h44m23.982080458s]"
time="2022-04-28T14:50:36Z" level=debug msg="faces: found no clusters to be merged [245.404163ms]"
time="2022-04-28T14:50:36Z" level=debug msg="faces: skipped clustering"
time="2022-04-28T14:50:36Z" level=debug msg="faces: found no new faces [77.358901ms]"
time="2022-04-28T14:50:36Z" level=debug msg="faces: found no unmatched markers"
time="2022-04-28T14:50:44Z" level=debug msg="faces: updated 0 markers, recognized 0 faces, 0 unknown [8.187022325s]"
time="2022-04-28T14:50:44Z" level=debug msg="index: updating counts"
time="2022-04-28T14:50:44Z" level=debug msg="counts: updated 0 places [102.806408ms]"
time="2022-04-28T14:50:44Z" level=debug msg="counts: updated 0 subjects [348.899356ms]"
time="2022-04-28T14:50:45Z" level=debug msg="counts: updated 0 labels [446.633104ms]"
time="2022-04-28T14:50:46Z" level=debug msg="index: found no photos without primary file [324.833524ms]"
time="2022-04-28T14:50:46Z" level=info msg="index: found no orphan files [184.496783ms]"
time="2022-04-28T14:50:48Z" level=debug msg="search: updated file index [867.313819ms]"
time="2022-04-28T14:50:48Z" level=debug msg="index: updating counts"
time="2022-04-28T14:50:49Z" level=debug msg="counts: updated 0 places [112.300731ms]"
time="2022-04-28T14:50:49Z" level=debug msg="counts: updated 0 subjects [396.137291ms]"
time="2022-04-28T14:50:49Z" level=debug msg="counts: updated 0 labels [393.100487ms]"
time="2022-04-28T14:50:49Z" level=debug msg="index: updating covers"
time="2022-04-28T14:50:49Z" level=debug msg="covers: updated 0 albums [39.950734ms]"
time="2022-04-28T14:50:50Z" level=debug msg="covers: updated 0 folders [150.710696ms]"
time="2022-04-28T14:50:50Z" level=debug msg="covers: updated 0 months [94.52815ms]"
time="2022-04-28T14:50:50Z" level=debug msg="covers: updated 0 labels [361.97886ms]"
time="2022-04-28T14:50:50Z" level=debug msg="covers: updated 0 subjects [173.919845ms]"
time="2022-04-28T14:50:50Z" level=info msg="indexed 322 files in 7h46m0.91184675s"
time="2022-04-28T14:50:50Z" level=info msg="closed database connection"
```


As suspected, It is indeed the `face` related commands which are invoked internally, as part of `index` command.

`time="2022-04-28T14:50:35Z" level=debug msg="faces: found no ambiguous subjects [7h44m23.982080458s]"`

This is the bottleneck. A whopping 7+hrs of checks for ambiguous subjects. I think it is fine to have this command. But please, could we not have this as part of `index`.

Do you guys agree to the analysis ? Or is it a peculiar problem with my setup ?

**Reply by srett** [reply]
Did you assign a lot of faces manually, i.e. not on the "new faces" page but on each individual photo? There's a performance bottleneck in the ResolveFaceCollisions function, as noted in #1872

**Reply by rickysarraf** [reply]
Maybe, yes. Facial recognition would report multiple faces under `/people/new`, where I'd tag them manually. Though I'm still not sure why this issue creeps up only when running the CLI `index` command. Today, I ran the (understandably) same thing from `WebUI => Start Indexing`, and it completed pretty quick.

```
photoprism_1  | time="2022-04-29T07:45:56Z" level=debug msg="moments: 513 albums saved as yaml files"                                                                                             
photoprism_1  | time="2022-04-29T07:45:56Z" level=info msg="indexing completed in 267 s"                                                                                                          
```

**Reply by srett** [reply]
> Maybe, yes. Facial recognition would report multiple faces under `/people/new`, where I'd tag them manually.

OK, that shouldn't be that bad, as you can see from the linked thread it was 10 minutes for me with 1500 tagged faces, 7 hours sounds like you tagged 10k+ faces. But you could check the number of rows in your faces table to be sure.

> Though I'm still not sure why this issue creeps up only when running the CLI index command. Today, I ran the (understandably) same thing from WebUI => Start Indexing, and it completed pretty quick.

It seems to only run the collision resolving immediately when indexing via CLI; otherwise, collision resolving is also run by the background worker asynchronously, every 15 minutes by default, so if you didn't change `PHOTOPRISM_WAKEUP_INTERVAL`, your CPU should also be hogged for 7 hours every 15 minutes. ;-)

**Reply by rickysarraf** [reply]
>OK, that shouldn't be that bad, as you can see from the linked thread it was 10 minutes for me with 1500 tagged faces, 7 hours sounds like you tagged 10k+ faces. But you could check the number of rows in your faces table to be sure.

Thank you. I ran command `select * from faces` and the result I get is: `3752 rows in set (0.051 sec)`.

>It seems to only run the collision resolving immediately when indexing via CLI; otherwise, collision resolving is also run by the background worker asynchronously, every 15 minutes by default, so if you didn't change PHOTOPRISM_WAKEUP_INTERVAL, your CPU should also be hogged for 7 hours every 15 minutes. ;-)

Yessss. The main process keeps hogging too. And now it all makes sense from what you've pointed out. I have now set `PHOTOPRISM_WAKEUP_INTERVAL` to `2700` seconds because it can only be set to a max of `3600` seconds if I want `facial recognition`.


I need to check the threads in Discussion and Issues to see what @lastzero hinted about. Somehow the db needs to be cleansed.

**Reply by rickysarraf** [reply]
@lastzero Would you agree to convert this into an issue ? I think it is common in today's day to have a large photo collection. And that users would tag faces randomly.

**Reply by lastzero** [reply]
You are welcome to open an issue when it's clear what the cause and likely solution is as we (and most others) have no such problems from what I see. 

It seems this is happening in a custom environment and so there can be many reasons why it is slow, not all have to do with our software or are supported by us... Like when indexing concurrently / with cron without an appropriate lock mechanism. We intentionally provide a background indexing example using ofelia for this reason.

**Reply by srett** [reply]
> You are welcome to open an issue when it's clear what the cause

I think it's what I outlined in #1872 – quadratic runtime. A few hundred rows in `faces` is fine, then it quickly spirals out of control. @rickysarraf has 3700 rows in the faces table, so, much worse than my test case even.

I was able to quickly reach this state through automation, which – fair enough – could be considered "unsupported use case". But in any case you'd be able to get there "by hand" if you patiently tag faces for an hour or so under "Show all new faces" (`/all?q=face%3Anew`), after already having tagged all clusters under `/people/new`.  
It appears that manually tagging individual faces – that could neither be clustered with already known faces, nor form a new unknown cluster – does not result in much further clustering afterwards either. Now whether this is a supported use case, I don't know, but I bet you'll have more such cases pop up over time, as I can totally see people's desire to have all the faces tagged properly and thus put in the time to tag individual ones manually. But it obviously takes a somewhat larger amount of photos, I guess many people run PhotoPrism with "only" a few thousand photos; you might not even have 3700 faces detected then, so never end up in this situation.

Should I take a stab at implementing the suggestion from #1872, adding a `checked_at` column to `faces`? It should work assuming the `updated_at` column is really always updated whenever a face cluster changes, gets merged, marker added/removed etc., but it might not result in the nicest code ever since I'm still not a Go dev, so it could end up at PoC quality level.

**Reply by lastzero** [reply]
We already implemented many code and documentation improvements for face recognition in our preview if that is the issue. Before adding more complexity and blocking other important tasks, we would like to release this first and see how it goes. Comments under the related GitHub issue on our roadmap suggest it is generally solved.

**Reply by lastzero** [reply]
We understand that not running the background worker often enough in combination with running multiple index jobs plus extensive manual tagging causes issues. Happy to attack such advanced use cases once we have enterprise customers paying for it.

**Comment by rickysarraf** [comment]
The performance also takes a hit with multiple processes spawned.

Here's the case:

```
 766922 rrs       20   0 4939548   1.6g  33428 S 140.2  20.5 422:05.71 photoprism                                                                                                                 
 792395 rrs       20   0 5344048   3.1g  44432 S 101.0  40.8 396:44.73 photoprism                                                                                                                 
```

2 jobs running with a substantial amount of resources consumed. One is the native process while the other one as part of my routine processing run.

```
rrs@lenovo:~$ ps aux | grep photoprism
rrs       763843  0.0  0.0   6872  2944 pts/2    S+   11:52   0:00 /bin/bash ./photoprism.sh
rrs       763844  0.1  0.5 347668 44620 pts/2    Sl+  11:52   0:40 /usr/bin/python3 /usr/bin/docker-compose -f stable-photoprism.yml up
rrs       763978  0.0  0.0   3944   192 ?        Ss   11:52   0:00 /bin/bash /scripts/entrypoint.sh /opt/photoprism/bin/photoprism start
rrs       766915  0.0  0.0   3944   124 ?        S    11:55   0:00 /bin/bash /scripts/entrypoint.sh /opt/photoprism/bin/photoprism start
rrs       766922  100 20.5 4939548 1657276 ?     Sl   11:55 424:42 /opt/photoprism/bin/photoprism start
rrs       792369  0.0  0.3 1191220 24796 pts/4   Sl+  12:34   0:03 docker exec photoprism_photoprism_1 photoprism index /Processed
rrs       792395  103 40.7 5344048 3292364 ?     Ssl  12:34 399:20 photoprism index /Processed
```

I don't know how best to tackle this scenario. In a strict scenario, a user shouldn't be allowed to have backend access. That will simply solve these `multiple processes` problem.

On the other hand, PhotoPrism doesn't mandate (yet) to not run commands from the CLI. So, as a user, I'm not doing anything wrong here.

In my opinion, resource intensive tasks (indexing, tagging, recognization etc) should have some sort of locking in place. I believe there's some locking in place already, but iirc it gets checked late.

**Reply by rickysarraf** [reply]
I think, other users that follow the recommended setup of setting cron like jobs through `ofelia`, will also run into such problems. So this problem shouldn't be unique to my setup.

**Reply by rickysarraf** [reply]
This is an erratic behavior as explained in [this comment](https://github.com/photoprism/photoprism/discussions/2287#discussioncomment-2660382).

**Reply by lastzero** [reply]
It's true we do not offer multi process / instance worker locking. Advanced users can use existing tools and methods to solve this problem. I'll take a look at other potential issues when I have a second. Had to take care of our site, terms, and privacy policy the last couple of days, so we hopefully don't get sued by a competitor or troll...
